### PR TITLE
Feature/vm enable fixedpoint

### PR DESCRIPTION
**Pull request by markos**
Initial FixedPoint VM enablement

### CODE REVIEW

File: libs/math/tests/math/all_tests/basic_math/standard_functions_test.cpp [hunk 1]
```diff
@@ -24,17 +24,55 @@
...
 using MyTypes = ::testing::Types<fetch::math::Tensor<float>, fetch::math::Tensor<double>,
                                  fetch::math::Tensor<fetch::fixed_point::FixedPoint<16, 16>>,
                                  fetch::math::Tensor<fetch::fixed_point::FixedPoint<32, 32>>>;
 
-TYPED_TEST_CASE(ClampTest, MyTypes);
+TYPED_TEST_CASE(StandardFunctionTests, MyTypes);
 
-TYPED_TEST(ClampTest, clamp_array_1D_test)
+template <typename TensorType>
+void RandomAssign(TensorType &tensor)
```
Comment (baykaner) [inline libs/math/tests/math/all_tests/basic_math/standard_functions_test.cpp:38]: there actually exist already some helper functions in tensor, such as:

Tensor<T, C>::UniformRandom
Tensor<T, C>::UniformRandomIntegers
Tensor<T, C>::FillUniformRandom
Tensor<T, C>::FillUniformRandomIntegers

At some point we should probably move a lot of this implementation out of tensor into free functions in the math library; but in any case the implementations are already there

File: libs/vectorise/include/vectorise/fixed_point/type_traits.hpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <type_traits>
+
+#include "core/byte_array/byte_array.hpp"
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/fixed_point/type_traits.hpp]: is this byte array include needed?

File: libs/vectorise/include/vectorise/fixed_point/type_traits.hpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#include "vectorise/fixed_point/fixed_point.hpp"
+
+namespace fetch {
+namespace math {
```
Comment (baykaner) [inline libs/vectorise/include/vectorise/fixed_point/type_traits.hpp:25]: should probably change the namespace to fetch::vectorise::meta, or fetch::vectorise::fixed_point::meta or something like that

File: libs/vm/include/vm/array.hpp [hunk 4]
```diff
@@ -293,6 +295,16 @@ inline Ptr<IArray> IArray::Construct(VM *vm, TypeId type_id, Args &&... args)
...
     return new Array<double>(vm, type_id, element_type_id, std::forward<Args>(args)...);
   }
+  case TypeIds::Fixed32:
+  {
+    return new Array<fixed_point::fp32_t>(vm, type_id, element_type_id,
+                                          std::forward<Args>(args)...);
+  }
+  case TypeIds::Fixed64:
+  {
+    return new Array<fixed_point::fp32_t>(vm, type_id, element_type_id,
```
Comment (baykaner) [inline libs/vm/include/vm/array.hpp]: shouldn't this be fixed_point::fp64_t ?

File: libs/vm/include/vm/map.hpp [hunk 5]
```diff
@@ -275,6 +284,14 @@ inline Ptr<IMap> outer(TypeId key_type_id, TypeId value_type_id, VM *vm, TypeId
   {
     return inner<double>(value_type_id, vm, type_id);
   }
+  // case TypeIds::Fixed32:
+  // {
```
Comment (baykaner) [inline libs/vm/include/vm/map.hpp]: leftover commented out code

File: libs/vm/include/vm/array.hpp (hunk 4 above)
Comment (markos) [reply libs/vm/include/vm/array.hpp]: fixed.

File: libs/vectorise/include/vectorise/fixed_point/type_traits.hpp (hunk 2 above)
Comment (markos) [reply libs/vectorise/include/vectorise/fixed_point/type_traits.hpp]: nope, fixed.

File: libs/vectorise/include/vectorise/fixed_point/type_traits.hpp (hunk 3 above)
Comment (markos) [reply libs/vectorise/include/vectorise/fixed_point/type_traits.hpp:25]: switched to fetch::fixed_point::meta

File: libs/math/tests/math/all_tests/basic_math/standard_functions_test.cpp (hunk 1 above)
Comment (markos) [reply libs/math/tests/math/all_tests/basic_math/standard_functions_test.cpp:38]: this was dragged in from elsewhere during a merge, not sure what I should do here to resolve this.

File: libs/vm/include/vm/map.hpp (hunk 5 above)
Comment (markos) [reply libs/vm/include/vm/map.hpp]: fixed

File: libs/vectorise/include/vectorise/fixed_point/type_traits.hpp (hunk 3 above)
Comment (markos) [reply libs/vectorise/include/vectorise/fixed_point/type_traits.hpp:25]: reverted to fetch::math::meta for (in)consistency reasons, ie fetch::fixed_point::meta::IsFixedPoint but fetch::math::meta::IsFloat.

File: libs/vm_modules/include/vm_modules/math/math.hpp [hunk 6]
```diff
@@ -40,6 +40,8 @@ static void CreateAbs(fetch::vm::Module &module)
   module.CreateFreeFunction<int32_t>("Abs", &Abs<int32_t>);
   module.CreateFreeFunction<float_t>("Abs", &Abs<float_t>);
   module.CreateFreeFunction<double_t>("Abs", &Abs<double_t>);
+  module.CreateFreeFunction<fixed_point::fp32_t>("Abs", &Abs<fixed_point::fp32_t>);
+  module.CreateFreeFunction<fixed_point::fp64_t>("Abs", &Abs<fixed_point::fp64_t>);
```
Comment (baykaner) [inline libs/vm_modules/include/vm_modules/math/math.hpp]: we should unify our modules_style. "Abs" here but "sqrt" and "pow" elsewhere.

Comment (markos) [reply libs/vm_modules/include/vm_modules/math/math.hpp]: agreed, I would go for lower-case everywhere, ie "abs".

**Review by baykaner — APPROVED** [review]
lgtm

**Pull request merged**